In [39]:
import pandas as pd
import numpy as np
import plotly.express as px

In [38]:
%pip install fairlearn

  Obtaining dependency information for fairlearn from https://files.pythonhosted.org/packages/b0/51/074c3e83d1254de673819e792326d74fc4b59cbc3820a02bc120b6edb854/fairlearn-0.9.0-py3-none-any.whl.metadata
   ---------------------------------------- 0.0/231.5 kB ? eta -:--:--
   -------------------------------------- - 225.3/231.5 kB 4.7 MB/s eta 0:00:01
   ---------------------------------------- 231.5/231.5 kB 4.8 MB/s eta 0:00:00
Note: you may need to restart the kernel to use updated packages.


In [99]:
df = pd.read_csv('heart.csv')
print(df.shape)
df.head(10)

(918, 12)


Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  MaxHR  \
0   40   M           ATA        140          289          0     Normal    172   
1   49   F           NAP        160          180          0     Normal    156   
2   37   M           ATA        130          283          0         ST     98   
3   48   F           ASY        138          214          0     Normal    108   
4   54   M           NAP        150          195          0     Normal    122   
5   39   M           NAP        120          339          0     Normal    170   
6   45   F           ATA        130          237          0     Normal    170   
7   54   M           ATA        110          208          0     Normal    142   
8   37   M           ASY        140          207          0     Normal    130   
9   48   F           ATA        120          284          0     Normal    120   

  ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
0              N      0.0       Up             0  
1              N      1.0     Flat             1  
2              N      0.0       Up             0  
3              Y      1.5     Flat             1  
4              N      0.0       Up             0  
5              N      0.0       Up             0  
6              N      0.0       Up             0  
7              N      0.0       Up             0  
8              Y      1.5     Flat             1  
9              N      0.0       Up             0

In [100]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 918 entries, 0 to 917
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Age             918 non-null    int64  
 1   Sex             918 non-null    object 
 2   ChestPainType   918 non-null    object 
 3   RestingBP       918 non-null    int64  
 4   Cholesterol     918 non-null    int64  
 5   FastingBS       918 non-null    int64  
 6   RestingECG      918 non-null    object 
 7   MaxHR           918 non-null    int64  
 8   ExerciseAngina  918 non-null    object 
 9   Oldpeak         918 non-null    float64
 10  ST_Slope        918 non-null    object 
 11  HeartDisease    918 non-null    int64  
dtypes: float64(1), int64(6), object(5)
memory usage: 86.2+ KB


In [101]:
df.describe().T

count        mean         std   min     25%    50%    75%    max
Age           918.0   53.510893    9.432617  28.0   47.00   54.0   60.0   77.0
RestingBP     918.0  132.396514   18.514154   0.0  120.00  130.0  140.0  200.0
Cholesterol   918.0  198.799564  109.384145   0.0  173.25  223.0  267.0  603.0
FastingBS     918.0    0.233115    0.423046   0.0    0.00    0.0    0.0    1.0
MaxHR         918.0  136.809368   25.460334  60.0  120.00  138.0  156.0  202.0
Oldpeak       918.0    0.887364    1.066570  -2.6    0.00    0.6    1.5    6.2
HeartDisease  918.0    0.553377    0.497414   0.0    0.00    1.0    1.0    1.0

RestingBP and Cholesterol have min 0
Change values from 0 to null for those two columns
#Need to discuss how to handle them 

In [102]:
df['RestingBP'] = df['RestingBP'].replace(0, np.nan)
df['Cholesterol'] = df['Cholesterol'].replace(0, np.nan)
df.isnull().sum() / df.shape[0]

Age               0.000000
Sex               0.000000
ChestPainType     0.000000
RestingBP         0.001089
Cholesterol       0.187364
FastingBS         0.000000
RestingECG        0.000000
MaxHR             0.000000
ExerciseAngina    0.000000
Oldpeak           0.000000
ST_Slope          0.000000
HeartDisease      0.000000
dtype: float64

In [103]:
df['HeartDisease'].value_counts()

1    508
0    410
Name: HeartDisease, dtype: int64

In [104]:
fig = px.pie(df, names='HeartDisease', title='Percentage for each outcome')
fig.show()

In [105]:
numerical= df.drop(['HeartDisease'], axis=1).select_dtypes('number').columns

categorical = df.select_dtypes('object').columns

print(f'Numerical Columns:  {df[numerical].columns}')
print('\n')
print(f'Categorical Columns: {df[categorical].columns}')

Numerical Columns:  Index(['Age', 'RestingBP', 'Cholesterol', 'FastingBS', 'MaxHR', 'Oldpeak'], dtype='object')


Categorical Columns: Index(['Sex', 'ChestPainType', 'RestingECG', 'ExerciseAngina', 'ST_Slope'], dtype='object')


In [106]:
df.groupby(['HeartDisease', 'Sex'])[['Age']].describe().T

HeartDisease           0                      1            
Sex                    F           M          F           M
Age count     143.000000  267.000000  50.000000  458.000000
    mean       51.202797   50.202247  56.180000   55.868996
    std         9.627981    9.344911   8.220656    8.788562
    min        30.000000   28.000000  33.000000   31.000000
    25%        44.000000   42.000000  51.000000   51.000000
    50%        51.000000   51.000000  58.000000   57.000000
    75%        57.000000   57.000000  62.000000   62.000000
    max        76.000000   75.000000  73.000000   77.000000

In [107]:
fig = px.box(df, y="Age", x="HeartDisease", points="all",  color="Sex",
             title="Distribution of age of database participants (stratified by outcome)")
fig.show()

In [108]:
fig = px.box(df, y="Age", x="HeartDisease", points="all", 
             title="Distribution of age of database participants (stratified by outcome)")
fig.show()

In [109]:
df_copy = df.copy()
df_copy['c'] = 1
df_copy.groupby(['HeartDisease', 'Sex']).agg('count')[['c']]

c
HeartDisease Sex     
0            F    143
             M    267
1            F     50
             M    458

In [110]:
fig = px.histogram(df_copy, x="Sex", y="c",
             color='HeartDisease', barmode='group',
             title="Distribution of outcome of database participants (stratified by sex)")
fig.show()

In [111]:
X= df.drop('HeartDisease', axis=1)
y= df['HeartDisease']

sex = df.pop("Sex") # Pop function drops and assigns at the same time
df.head(10)

Age ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  MaxHR  \
0   40           ATA      140.0        289.0          0     Normal    172   
1   49           NAP      160.0        180.0          0     Normal    156   
2   37           ATA      130.0        283.0          0         ST     98   
3   48           ASY      138.0        214.0          0     Normal    108   
4   54           NAP      150.0        195.0          0     Normal    122   
5   39           NAP      120.0        339.0          0     Normal    170   
6   45           ATA      130.0        237.0          0     Normal    170   
7   54           ATA      110.0        208.0          0     Normal    142   
8   37           ASY      140.0        207.0          0     Normal    130   
9   48           ATA      120.0        284.0          0     Normal    120   

  ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
0              N      0.0       Up             0  
1              N      1.0     Flat             1  
2              N      0.0       Up             0  
3              Y      1.5     Flat             1  
4              N      0.0       Up             0  
5              N      0.0       Up             0  
6              N      0.0       Up             0  
7              N      0.0       Up             0  
8              Y      1.5     Flat             1  
9              N      0.0       Up             0

In [112]:
numeric_features = ["Age", "RestingBP", "Cholesterol", "Oldpeak", "RestingBP", "MaxHR", "Oldpeak"]
X[numeric_features]


Age  RestingBP  Cholesterol  Oldpeak  RestingBP  MaxHR  Oldpeak
0     40      140.0        289.0      0.0      140.0    172      0.0
1     49      160.0        180.0      1.0      160.0    156      1.0
2     37      130.0        283.0      0.0      130.0     98      0.0
3     48      138.0        214.0      1.5      138.0    108      1.5
4     54      150.0        195.0      0.0      150.0    122      0.0
..   ...        ...          ...      ...        ...    ...      ...
913   45      110.0        264.0      1.2      110.0    132      1.2
914   68      144.0        193.0      3.4      144.0    141      3.4
915   57      130.0        131.0      1.2      130.0    115      1.2
916   57      130.0        236.0      0.0      130.0    174      0.0
917   38      138.0        175.0      0.0      138.0    173      0.0

[918 rows x 7 columns]

In [113]:
data = X[numeric_features]
X_num = pd.DataFrame(data, columns=numeric_features)
X_num

Age  RestingBP  Cholesterol  Oldpeak  RestingBP  MaxHR  Oldpeak
0     40      140.0        289.0      0.0      140.0    172      0.0
1     49      160.0        180.0      1.0      160.0    156      1.0
2     37      130.0        283.0      0.0      130.0     98      0.0
3     48      138.0        214.0      1.5      138.0    108      1.5
4     54      150.0        195.0      0.0      150.0    122      0.0
..   ...        ...          ...      ...        ...    ...      ...
913   45      110.0        264.0      1.2      110.0    132      1.2
914   68      144.0        193.0      3.4      144.0    141      3.4
915   57      130.0        131.0      1.2      130.0    115      1.2
916   57      130.0        236.0      0.0      130.0    174      0.0
917   38      138.0        175.0      0.0      138.0    173      0.0

[918 rows x 7 columns]

In [114]:
categorical_features = ["ChestPainType", "RestingECG", "ST_Slope"]
X[categorical_features]

ChestPainType RestingECG ST_Slope
0             ATA     Normal       Up
1             NAP     Normal     Flat
2             ATA         ST       Up
3             ASY     Normal     Flat
4             NAP     Normal       Up
..            ...        ...      ...
913            TA     Normal     Flat
914           ASY     Normal     Flat
915           ASY     Normal     Flat
916           ATA        LVH     Flat
917           NAP     Normal       Up

[918 rows x 3 columns]

In [115]:
binary_features = ["Sex", "FastingBS", "ExerciseAngina"]
X[binary_features]

Sex  FastingBS ExerciseAngina
0     M          0              N
1     F          0              N
2     M          0              N
3     F          0              Y
4     M          0              N
..   ..        ...            ...
913   M          0              N
914   M          1              N
915   M          0              Y
916   F          0              N
917   M          0              N

[918 rows x 3 columns]

In [116]:
from sklearn.preprocessing import OrdinalEncoder

binary_transformer = OrdinalEncoder()

data = binary_transformer.fit_transform(X[binary_features])
X_bin = pd.DataFrame(data, columns=binary_features)
X_bin

Sex  FastingBS  ExerciseAngina
0    1.0        0.0             0.0
1    0.0        0.0             0.0
2    1.0        0.0             0.0
3    0.0        0.0             1.0
4    1.0        0.0             0.0
..   ...        ...             ...
913  1.0        0.0             0.0
914  1.0        1.0             0.0
915  1.0        0.0             1.0
916  0.0        0.0             0.0
917  1.0        0.0             0.0

[918 rows x 3 columns]

In [117]:
from sklearn.preprocessing import OneHotEncoder
categorical_transformer = OneHotEncoder(handle_unknown="ignore")
data = categorical_transformer.fit_transform(X[categorical_features]).toarray()
cols = categorical_transformer.get_feature_names_out()
X_cat = pd.DataFrame(data, columns=cols)
X_cat

ChestPainType_ASY  ChestPainType_ATA  ChestPainType_NAP  \
0                  0.0                1.0                0.0   
1                  0.0                0.0                1.0   
2                  0.0                1.0                0.0   
3                  1.0                0.0                0.0   
4                  0.0                0.0                1.0   
..                 ...                ...                ...   
913                0.0                0.0                0.0   
914                1.0                0.0                0.0   
915                1.0                0.0                0.0   
916                0.0                1.0                0.0   
917                0.0                0.0                1.0   

     ChestPainType_TA  RestingECG_LVH  RestingECG_Normal  RestingECG_ST  \
0                 0.0             0.0                1.0            0.0   
1                 0.0             0.0                1.0            0.0   
2                 0.0             0.0                0.0            1.0   
3                 0.0             0.0                1.0            0.0   
4                 0.0             0.0                1.0            0.0   
..                ...             ...                ...            ...   
913               1.0             0.0                1.0            0.0   
914               0.0             0.0                1.0            0.0   
915               0.0             0.0                1.0            0.0   
916               0.0             1.0                0.0            0.0   
917               0.0             0.0                1.0            0.0   

     ST_Slope_Down  ST_Slope_Flat  ST_Slope_Up  
0              0.0            0.0          1.0  
1              0.0            1.0          0.0  
2              0.0            0.0          1.0  
3              0.0            1.0          0.0  
4              0.0            0.0          1.0  
..             ...            ...          ...  
913            0.0            1.0          0.0  
914            0.0            1.0          0.0  
915            0.0            1.0          0.0  
916            0.0            1.0          0.0  
917            0.0            0.0          1.0  

[918 rows x 10 columns]

In [118]:
X_data = pd.concat([X_cat, X_bin, X_num], axis = 1)
X_data

ChestPainType_ASY  ChestPainType_ATA  ChestPainType_NAP  \
0                  0.0                1.0                0.0   
1                  0.0                0.0                1.0   
2                  0.0                1.0                0.0   
3                  1.0                0.0                0.0   
4                  0.0                0.0                1.0   
..                 ...                ...                ...   
913                0.0                0.0                0.0   
914                1.0                0.0                0.0   
915                1.0                0.0                0.0   
916                0.0                1.0                0.0   
917                0.0                0.0                1.0   

     ChestPainType_TA  RestingECG_LVH  RestingECG_Normal  RestingECG_ST  \
0                 0.0             0.0                1.0            0.0   
1                 0.0             0.0                1.0            0.0   
2                 0.0             0.0                0.0            1.0   
3                 0.0             0.0                1.0            0.0   
4                 0.0             0.0                1.0            0.0   
..                ...             ...                ...            ...   
913               1.0             0.0                1.0            0.0   
914               0.0             0.0                1.0            0.0   
915               0.0             0.0                1.0            0.0   
916               0.0             1.0                0.0            0.0   
917               0.0             0.0                1.0            0.0   

     ST_Slope_Down  ST_Slope_Flat  ST_Slope_Up  Sex  FastingBS  \
0              0.0            0.0          1.0  1.0        0.0   
1              0.0            1.0          0.0  0.0        0.0   
2              0.0            0.0          1.0  1.0        0.0   
3              0.0            1.0          0.0  0.0        0.0   
4              0.0            0.0          1.0  1.0        0.0   
..             ...            ...          ...  ...        ...   
913            0.0            1.0          0.0  1.0        0.0   
914            0.0            1.0          0.0  1.0        1.0   
915            0.0            1.0          0.0  1.0        0.0   
916            0.0            1.0          0.0  0.0        0.0   
917            0.0            0.0          1.0  1.0        0.0   

     ExerciseAngina  Age  RestingBP  Cholesterol  Oldpeak  RestingBP  MaxHR  \
0               0.0   40      140.0        289.0      0.0      140.0    172   
1               0.0   49      160.0        180.0      1.0      160.0    156   
2               0.0   37      130.0        283.0      0.0      130.0     98   
3               1.0   48      138.0        214.0      1.5      138.0    108   
4               0.0   54      150.0        195.0      0.0      150.0    122   
..              ...  ...        ...          ...      ...        ...    ...   
913             0.0   45      110.0        264.0      1.2      110.0    132   
914             0.0   68      144.0        193.0      3.4      144.0    141   
915             1.0   57      130.0        131.0      1.2      130.0    115   
916             0.0   57      130.0        236.0      0.0      130.0    174   
917             0.0   38      138.0        175.0      0.0      138.0    173   

     Oldpeak  
0        0.0  
1        1.0  
2        0.0  
3        1.5  
4        0.0  
..       ...  
913      1.2  
914      3.4  
915      1.2  
916      0.0  
917      0.0  

[918 rows x 20 columns]

In [136]:
from  sklearn.tree import DecisionTreeClassifier

y_true = y
cf = DecisionTreeClassifier(min_samples_leaf=10, max_depth=4) #parameters have not been tuned
cf.fit(X_data, y)
y_pred = cf.predict(X_data) 

In [137]:
from fairlearn.metrics import MetricFrame
from sklearn.metrics import accuracy_score

gm = MetricFrame(accuracy_score, y_true, y_pred, sensitive_features=sex)
print(gm.overall)

0.8246187363834423


C:\Users\Yerin\anaconda3\Lib\site-packages\fairlearn\metrics\_metric_frame.py:77: FutureWarning:

You have provided 'metrics', 'y_true', 'y_pred' as positional arguments. Please pass them as keyword arguments. From version 0.10.0 passing them as positional arguments will result in an error.



In [138]:
print(gm.by_group)

Sex
F    0.813472
M    0.827586
Name: accuracy_score, dtype: float64


In [140]:
from fairlearn.metrics import selection_rate
sr = MetricFrame(selection_rate, y_true, y_pred, sensitive_features=sex)
sr.overall

C:\Users\Yerin\anaconda3\Lib\site-packages\fairlearn\metrics\_metric_frame.py:77: FutureWarning:

You have provided 'metrics', 'y_true', 'y_pred' as positional arguments. Please pass them as keyword arguments. From version 0.10.0 passing them as positional arguments will result in an error.



0.5261437908496732

In [141]:
sr.by_group

Sex
F    0.362694
M    0.569655
Name: selection_rate, dtype: float64

In [142]:
from fairlearn.metrics import demographic_parity_difference, equalized_odds_difference, demographic_parity_ratio

dpd = demographic_parity_difference(y_true, y_pred, sensitive_features = sex)
eod = equalized_odds_difference(y_true, y_pred, sensitive_features = sex)
dpr = demographic_parity_ratio(y_true, y_pred, sensitive_features = sex)

print("Demographic parity difference: {}".format(round(dpd, 2)))
print("Equalized odds difference: {}".format(round(eod, 2)))
print("Demographic parity ratio: {}".format(round(dpr, 2)))

Demographic parity difference: 0.21
Equalized odds difference: 0.05
Demographic parity ratio: 0.64


1) Demographic difference in parity: if the absolute value is less than 0.1, then the model can be considered fair.
2) Balanced odds difference: if the absolute value is less than 0.1, then the model can be considered fair.
3) The difference is in equal opportunities: if the absolute value is less than 0.1, then the model can be considered fair.
4) Demographic parity coefficient: the fairness of this indicator ranges from 0.8 to 1.25.

In our case, the difference in demographic parity (0.21), demographic parity ratio (0.64)indicate unfairness, while the difference in equalization of chances (0.05) does not indicate unfairness.

In [143]:
from fairlearn.reductions import ExponentiatedGradient, DemographicParity, EqualizedOdds

classifier = DecisionTreeClassifier(min_samples_leaf=10, max_depth=4)
dp = DemographicParity()
reduction = ExponentiatedGradient(classifier, dp)

reduction.fit(X_data, y, sensitive_features=sex)
prediction_dp = reduction.predict(X_data)

In [146]:
sr_mitigated = MetricFrame(selection_rate, y_true, prediction_dp, sensitive_features=sex)
print(sr_mitigated.overall)

0.6078431372549019


C:\Users\Yerin\anaconda3\Lib\site-packages\fairlearn\metrics\_metric_frame.py:77: FutureWarning:

You have provided 'metrics', 'y_true', 'y_pred' as positional arguments. Please pass them as keyword arguments. From version 0.10.0 passing them as positional arguments will result in an error.



In [147]:
print(sr_mitigated.by_group)

Sex
F    0.601036
M    0.609655
Name: selection_rate, dtype: float64


In [150]:
gm = MetricFrame(accuracy_score, y_true, prediction_dp, sensitive_features=sex)
print(gm.overall)

0.8148148148148148


C:\Users\Yerin\anaconda3\Lib\site-packages\fairlearn\metrics\_metric_frame.py:77: FutureWarning:

You have provided 'metrics', 'y_true', 'y_pred' as positional arguments. Please pass them as keyword arguments. From version 0.10.0 passing them as positional arguments will result in an error.



In [149]:
dpd = demographic_parity_difference(y_true, prediction_dp, sensitive_features = sex)
eod = equalized_odds_difference(y_true, prediction_dp, sensitive_features = sex)
dpr = demographic_parity_ratio(y_true, prediction_dp, sensitive_features = sex)

print("Demographic parity difference: {}".format(round(dpd, 2)))
print("Equalized odds difference: {}".format(round(eod, 2)))
print("Demographic parity ratio: {}".format(round(dpr, 2)))

Demographic parity difference: 0.01
Equalized odds difference: 0.34
Demographic parity ratio: 0.99


1) Demographic difference in parity: if the absolute value is less than 0.1, then the model can be considered fair.
2) Balanced odds difference: if the absolute value is less than 0.1, then the model can be considered fair.
3) The difference is in equal opportunities: if the absolute value is less than 0.1, then the model can be considered fair.
4) Demographic parity coefficient: the fairness of this indicator ranges from 0.8 to 1.25.

In our case, the difference in demographic parity (0.01), demographic parity ratio (0.99) does not indicate unfairness, while the difference in equalization of chances (0.34) indicates unfairness.